### ISSUE TITLE
Chapter 9 comments

### ISSUE THREAD

**Issue by RaphaelS1** [mlr-org/mlr3book#442]
## Style

* [x] Suggest briefly distinguishing between implicit and explicit here
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L18
* [x] In this section I'd suggest adding a diagram at the end to help reader understanding.
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L91
* [x] In this section I'd suggest giving a brief overview to all future methods and assume readers no nothing.
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L91

Content
* [ ] Suggest using chapter to describe issues you’ve had with R6 and how you’ve solved them, e.g. serialisation and backend S3 methods, syntactic sugar, etc. This chapter is the space to write as much detail as you want about things you care about. ‘Technical’ does not need to only refer to software, what about long discussions we’ve had around benchmarking, optimal tuning, etc.
* [x] But why would someone want to do this (pick different strategies)? Can't just show possibilities in mlr, also need to guide the reader a bit. This might be a good example of a diagram being useful.
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L144
* [x] Intro is confusing, I don’t understand the point of classif.debug how is it used to debug other learners?
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L177
* [x] Really important section, could do with a bit more detail.
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L219
* [x] Printing this takes up space and isn't useful
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L300
* [x] Output not useful, doesn't show final scores or the errors column. Subset columns.
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L316
* [x] How is this "further investigating errors"? What do I do with this table? Nowhere in this section have I been shown an actual error I can deal with
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L324
* [x] If you're making comments like "long story short" then you're example needs to be changed. Readers don't want a long story. Example is overly complex
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L330-L338 
* [x] Difference between these hard to understand, rewrite slightly.
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L368-L369
* [ ]  Be careful not to duplicate content in earlier chapters. Also be careful to not just write a book for paradox (though I appreciate it’s an mlr-org package). Don’t duplicate R6 documentation.
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L510
* [x] https://github.com/mlr-org/mlr3book/issues/250

## Structure

* [ ] I suggest reordering chapter by most common use-cases. So: hyperparameters → logging → debugging → parallelization → out-of-memory data.
* [ ] I’m unsure if the subsections should really be separated, or maybe 6.2.2 should actually go first and then in 6.2.1 you explain different methods of encapsulation. Currently a bit hard to follow, especially as fallbacks are probably more common use-case (at least once you're interested in this level of technical detail)
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L175
* [x] Section would probably be clearer without further headings and instead just one worked examples, I've made similar comments in #440
https://github.com/mlr-org/mlr3book/blob/ac3f3b95eee0f4d1556734d83d05860596ad4c16/book/technical.qmd#L372

**Comment by mllg** [comment]
Thanks for your comments. I hope I will find some time to work on the book next week.

**Comment by RaphaelS1** [comment]
Thanks!

**Comment by sebffischer** [comment]
I guess this also falls under the category "issues you had with R6": What would be great is some information about `leanification` as this also has some bearings on how loading / reading mlr3 objects behave when a different mlr3 version is installed.

**Comment by sebffischer** [comment]
@mllg I will do the paradox chapter (at least this is what I agreed with Marc).

**Comment by RaphaelS1** [comment]
Thanks @sebffischer these are all great additions

**Comment by mllg** [comment]
@RaphaelS1 can you elaborate what kind of diagrams you would like to see in the section on explicit parallelization?

**Comment by mllg** [comment]
Would this mermaid diagram help (this currently does not render in our CI)?


```{mermaid}
%%| label: fig-parallel
%%| fig-cap: "Principle of parallel computing: A task consisting of multiple subtasks is spawned on 3 workers. The results are collected (and usually combined) in a second step."
graph LR
    M[fa:fa-server Main]
    S{Spawn}
    C{Collect}

    M --> S
    S -->|Task 1| W1[fa:fa-microchip Worker 1]
    S -->|Task 2| W2[fa:fa-microchip Worker 2]
    S -->|Task 3| W3[fa:fa-microchip Worker 3]
    W1 -->|Result 1| C
    W2 -->|Result 2| C
    W3 -->|Result 3| C
```

**Comment by RaphaelS1** [comment]
What about something like:

```
flowchart LR
    
    subgraph SG1[fa:fa-server Main]
        S[(Task)]
    end
    subgraph SG2[fa:fa-microchip Worker 1]
        T1[(Fold 1)]
    end
    subgraph SG3[fa:fa-microchip Worker 2]
        T2[(Fold 2)]
    end
    subgraph SG4[fa:fa-microchip Worker 1]
        T11[(Fold 1.1)]
    end
    subgraph SG5[fa:fa-microchip Worker 3]
        T12[(Fold 1.2)]
    end
    subgraph SG6[fa:fa-microchip Worker 2]
        T21[(Fold 2.1)]
    end
    subgraph SG7[fa:fa-microchip Worker 4]
        T22[(Fold 2.2)]
    end
    subgraph SG8[fa:fa-microchip Worker 1]
        C1[Aggregate]
    end
    subgraph SG9[fa:fa-microchip Worker 2]
        C2[Aggregate]
    end
    subgraph SG10[fa:fa-server Main]
        direction LR
        C[Aggregate] --> R([Results])
    end

    SG1 -->|Resample| SG2 & SG3
    SG2 -->|Resample| SG4 & SG5 -->|Collect| SG8
    SG3 -->|Resample| SG6 & SG7 -->|Collect| SG9

    SG8 & SG9 -->|Collect| SG10
    
    classDef sg fill:white,stroke:#D3D3D3,stroke-width:1px;
    class SG1,SG2,SG3,SG4,SG5,SG6,SG7,SG8,SG9,SG10 sg;
    
    classDef default fill:white,stroke:#333,stroke-width:2px height:80px;
```

![Screenshot 2022-11-18 at 13 00 50](https://user-images.githubusercontent.com/25639974/202711102-413b2462-8194-4f57-be1c-c3d3b1f900f3.png)

**Comment by mllg** [comment]
I think I've addressed most of your points in #450. I guess there will be a 2nd round by Bernd or Lars?

**Comment by RaphaelS1** [comment]
Thanks!! Before tagging in Bernd can you just clarify why the points not checked have been left out? Copying here for ease:

* [ ] Suggest using chapter to describe issues you’ve had with R6 and how you’ve solved them, e.g. serialisation (and leanification) and backend S3 methods, syntactic sugar, etc. This chapter is the space to write as much detail as you want about things you care about. ‘Technical’ does not need to only refer to software, what about long discussions we’ve had around benchmarking, optimal tuning, etc. 
* [ ] Be careful not to duplicate content in earlier chapters. Also be careful to not just write a book for paradox (though I appreciate it’s an mlr-org package). Don’t duplicate R6 documentation.
* [ ] #250 
* [ ]  I suggest reordering chapter by most common use-cases. So: hyperparameters → logging → debugging → parallelization → out-of-memory data.
* [ ]  I’m unsure if the subsections should really be separated, or maybe 6.2.2 should actually go first and then in 6.2.1 you explain different methods of encapsulation. Currently a bit hard to follow, especially as fallbacks are probably more common use-case (at least once you're interested in this level of technical detail)
* [ ] The option of doing parallel prediction could be added to the parallelization chapter
* [ ] but maybe have some explanation of multithreading + multiprocessing?
* [ ] In this section I'd suggest giving a brief overview to all `future` methods and assume readers no nothing.
* [ ]  Printing this takes up space and isn't useful
* [ ] If you're making comments like "long story short" then you're example needs to be changed. Readers don't want a long story. Example is overly complex -> Example is still overly complex, a real-world example is more useful, sorry but even I still struggle to understand this having read it 3 times
* [ ] 



Changes are really great, thanks a lot, but two major things to address before Bernd's read:

1. [ ] Please add a longer introduction, see Optimization chapter version in https://github.com/mlr-org/mlr3book/pull/457 here as an example
2. [ ] Please fill out the chapter abstract

**Comment by RaphaelS1** [comment]
Copying from #430 

> Two points from the other issue (https://github.com/mlr-org/mlr3book/issues/442):

> * [ ]    parallel predict
> * [ ] R6 quirks (?) ... I personally don't think this should be covered in the book though

**Comment by RaphaelS1** [comment]
Please also

* [ ] Use `index` to add terms to add to the index
* [ ] Use `link` to create accessible links

**Comment by RaphaelS1** [comment]
Sorry hopefully just one more thing:

* [ ] Change `spam` -> `sonar`

**Comment by larskotthoff** [comment]
- [ ] In the abstract, I would outline what the chapter covers rather than warning people that some things may be too advanced.
- [ ] "Unfortunately, we let out the more technical details yet, but they are utterly important for conducting applied ML experiments." -- this sentence is weird. Sounds like somebody who had no control over the earlier chapters wrote this one and now has to address everything that's wrong with the other content. Rephrase to something like "so far, we have covered ML concepts without going into a lot of technical detail, which can be important for more advanced uses of mlr3".
- [ ] "adding new classes to mlr3" is covered in a different chapter and shouldn't be mentioned here.
- [ ] 9.1 forks -> processes
- [ ] 9.1 where "embarrassingly parallel" is mentioned I would also mention data parallelism, i.e. the reason these operations are so easy to parallelize is that the same operation is performed on different parts of the data.
- [ ] 9.1 I would remove the "abstract" example at the end -- it illustrates general parallelization concepts more than anything specific to mlr3 (especially as different libraries are used in mlr3). If there's an example that uses mlr3 to illustrate the same concept I would use that.
- [ ] 9.1 explanations for batchtools::binpack and lpt seem to be reversed and links to the help pages are broken.
- [ ] 9.1.2 show the output of `future::plan()` and `diff`.
- [ ] 9.1.2 shows an example with "multisession", but then recommends "multicore" in the text -- change example or explain why "multisession" is shown.
- [ ] 9.1.3 would be good to have an example of benchmark parallelization.
- [ ] 9.1.4 mention that whether multiple HP configs can be evaluated also depends on the HPO method.
- [ ] 9.1.4 what would happen if you simply run `future::plan("multisession")` for this example? Seems like this should create 10 threads and optimally parallelize (and also be much easier).
- [ ] 9.1.4 FIXME at the end of section.
- [ ] 9.1.6 leaves the reader hanging -- do I need to worry about reproducibility or not? How do I make parallel experiments with mlr3 reproducible?
- [ ] 9.2 what do HPs `x` and `iter` do for "classif.debug"?
- [ ] 9.2.1 and 9.2.2 are also covered in the HPO chapter -- I think that this is more appropriate here, could only mention in HPO chapter.
- [ ] 9.4.1 suppress output from attaching libraries.
- [ ] 9.5 should probably be moved to chapter 11.
- [ ] 9.5 explain indicator function.
- [ ] 9.6 list the things that were covered.
- [ ] 9.7 I would use the parallelization of a benchmark experiment as an exercise, i.e. something that directly uses mlr3.

**Comment by larskotthoff** [comment]
ping @mllg

**Comment by mllg** [comment]
> [ ] Change `spam` -> `sonar`

@RaphaelS1 is this still up to date? I've picked spam because the runtime of the training is measurable.

**Comment by mllg** [comment]
> 9.1.4 what would happen if you simply run future::plan("multisession") for this example? Seems like this should create 10 threads and optimally parallelize (and also be much easier).

@be-marc maybe it is too late, but can you double-check what happens here? Do we need to pick a different HPO algorithm or batch size for this toy example to work?

**Comment by mllg** [comment]
> "adding new classes to mlr3" is covered in a different chapter and shouldn't be mentioned here.

AFAIK the extending stuff should go into vignettes of respective packages and we only keep extending measures here as an example.

> 9.1 where "embarrassingly parallel" is mentioned I would also mention data parallelism, i.e. the reason these operations are so easy to parallelize is that the same operation is performed on different parts of the data.


I am not an expert in this field, but after reading the Wikipedia page, I have my doubts that resampling in the strict sense is data parallel.
We do not partition the data into equally sized (and disjoint) sets, but operate on arbitrary subsets.
If you are sure, I can add a sentence here though.


 > 9.1 I would remove the "abstract" example at the end -- it illustrates general parallelization concepts more than anything specific to mlr3 (especially as different libraries are used in mlr3). If there's an example that uses mlr3 to illustrate the same concept I would use that.

I really would like to keep it, even though you are right that it does not directly translates to something in mlr3.
Instead of removing it, I would like to add 2-3 sentences later and show how to control chunking via future so that we have the bridge to mlr3. This is what the FIXME was for 🙈

**Comment by mllg** [comment]
Here is the new PR: #576

**Comment by larskotthoff** [comment]
> > 9.1 where "embarrassingly parallel" is mentioned I would also mention data parallelism, i.e. the reason these operations are so easy to parallelize is that the same operation is performed on different parts of the data.
> 
> I am not an expert in this field, but after reading the Wikipedia page, I have my doubts that resampling in the strict sense is data parallel. We do not partition the data into equally sized (and disjoint) sets, but operate on arbitrary subsets. If you are sure, I can add a sentence here though.

AFAIK data parallelism is simply running the same algorithm on multiple parts of the data, which we certainly do. I would add a sentence here.
